### PR TITLE
feat: subscription, hostingplatform and dnszone components

### PR DESCRIPTION
**Pull request by ml019**
## Intent of Change
- New feature (non-breaking change which adds functionality)

## Description
Add support for new `subscription`, `hostingplatform` and `dnszone` components.

## Motivation and Context
Part of [ADR0007](https://github.com/hamlet-io/architectural-decision-log/blob/main/adr/0007-placement-support.md) implementation.

- Add shared definition of these new components
- Add preliminary changes to support `Locations`
  - accessor methods for location information on components and occurrences
  - ability to capture the errors generated by `getCompositeObject()` so alternate actions to generating fatal errors can be supported, mainly so the context of the errors can be better explained
  - More extensive commenting of the process for attribute extension and some refactoring of the implementation
- new component attributes support import of externally defined resources in readiness for the introduction of `Locations`.
- Support for an explicit Type attribute on components rather than relying on the name of a child attribute
  - makes config more explicit
  - permits a flatter JSON file
  - in line with similar changes to links with the removal of subcomponent specific attributes

Subsequent PRs will add explicit support for Locations.
## How Has This Been Tested?
Local template generation

### Prerequisite PRs:

- None

### Dependent PRs:

- None

### Consumer Actions:

- None

### CODE REVIEW

File: engine/component.ftl [hunk 1]
```diff
@@ -241,24 +271,80 @@
...
+                "Children" : [
+                    {
+                        "Names" : "Mandatory",
+                        "Description" : "Is location always required?",
+                        "Types" : BOOLEAN_TYPE,
+                        "Default" : true
+                    },
+                    {
+                        "Names" : "TargetComponentTypes",
+                        "Description" : "Enable at rest encryption",
```
Comment (roleyfoley) [inline engine/component.ftl:289]: Is that description right?

File: engine/setContext.ftl [hunk 2]
```diff
@@ -3,6 +3,8 @@
 [#-- Shared Provider Configurations --]
 [@includeSharedComponentConfiguration component="shared" /]
 [@includeSharedComponentConfiguration component="baseline" /]
+[@includeSharedComponentConfiguration component="s3" /]
+[@includeSharedComponentConfiguration component="ec2" /]
```
Comment (roleyfoley) [inline engine/setContext.ftl:7]: Do we need to include new providers? I thought we were trying to drop this?

**Review by roleyfoley — CHANGES_REQUESTED** [review]

File: engine/setContext.ftl (hunk 2 above)
Comment (ml019) [reply engine/setContext.ftl:7]: Its temporary - they disappear in the districts PR.

File: engine/component.ftl (hunk 1 above)
Comment (ml019) [reply engine/component.ftl:289]: Cut/paste error - I'll fix

**Review by roleyfoley — APPROVED** [review]

**Pull request merged**